services/node_listener_test.go

```diff
@@ -0,0 +1,346 @@
+		{"runlog matching address", "runlog", sharedAddr, sharedAddr, 1, cltest.StringToRunLogData(`{"value":"100"}`)},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			nl, cleanup := cltest.NewNodeListener()
+			defer cleanup()
+			store := nl.Store
+			cltest.MockEthOnStore(store)
+			assert.Nil(t, nl.HeadTracker.Start())
+			assert.Nil(t, nl.Start())
+
+			eth := cltest.MockEthOnStore(store)
+			logChan := make(chan types.Log, 1)
+			eth.RegisterSubscription("logs", logChan)
+
+			j := cltest.NewJob()
+			initr := models.Initiator{Type: test.initType}
+			if !utils.IsEmptyAddress(test.initrAddr) {
+				initr.Address = test.initrAddr
+			}
+			j.Initiators = []models.Initiator{initr}
+			assert.Nil(t, store.SaveJob(&j))
+
+			nl.AddJob(j)
+
+			logChan <- types.Log{
+				Address: test.logAddr,
+				Data:    test.data,
+				Topics: []common.Hash{
+					services.RunLogTopic,
+					common.StringToHash("requestID"),
+					common.StringToHash(j.ID),
+				},
+			}
+
+			cltest.WaitForRuns(t, j, store, test.wantCount)
+
+			eth.EnsureAllCalled(t)
+		})
+	}
+}
+
+func TestNodeListener_newHeadsNotification(t *testing.T) {
+	t.Parallel()
+
+	app, cleanup := cltest.NewApplicationWithKeyStore()
+	defer cleanup()
+	store := app.Store
+
+	ethMock := app.MockEthClient()
+	nhChan := ethMock.RegisterNewHeads()
+	ethMock.Register("eth_getTransactionReceipt", strpkg.TxReceipt{})
+	sentAt := uint64(23456)
+	confirmationAt := sentAt + 1
+	ethMock.Register("eth_blockNumber", utils.Uint64ToHex(confirmationAt))
+
+	app.Start()
+
+	j := models.NewJob()
+	j.Tasks = []models.Task{cltest.NewTask("ethtx")}
+	assert.Nil(t, store.SaveJob(&j))
+
+	tx := cltest.CreateTxAndAttempt(store, cltest.NewAddress(), sentAt)
+	txas, err := store.AttemptsFor(tx.ID)
+	assert.Nil(t, err)
+	txa := txas[0]
+
+	jr := j.NewRun()
+	tr := jr.TaskRuns[0]
+	result := cltest.RunResultWithValue(txa.Hash.String())
+	tr.Result = result.MarkPending()
+	tr.Status = models.StatusPending
+	jr.TaskRuns[0] = tr
+	jr.Status = models.StatusPending
+	assert.Nil(t, store.Save(&jr))
+
+	blockNumber := cltest.BigHexInt(1)
+	nhChan <- models.BlockHeader{Number: blockNumber}
+
+	ethMock.EnsureAllCalled(t)
+	assert.Equal(t, blockNumber, app.NodeListener.HeadTracker.Get().Number)
+}
+
+func TestHeadTracker_New(t *testing.T) {
+	t.Parallel()
+
+	store, cleanup := cltest.NewStore()
+	defer cleanup()
+	cltest.MockEthOnStore(store)
+	assert.Nil(t, store.Save(models.NewIndexableBlockNumber(big.NewInt(1))))
+	last := models.NewIndexableBlockNumber(big.NewInt(0x10))
+	assert.Nil(t, store.Save(last))
+	assert.Nil(t, store.Save(models.NewIndexableBlockNumber(big.NewInt(0xf))))
+
+	ht := services.NewHeadTracker(store)
+	assert.Nil(t, ht.Start())
+	assert.Equal(t, last.Number, ht.Get().Number)
+}
+
+func TestHeadTracker_Get(t *testing.T) {
+	t.Parallel()
+
+	start := models.NewIndexableBlockNumber(big.NewInt(5))
```

Comment:
I know this is not new code just moved to a new file, so this suggestion is out of scope of the current PR, but I'm going to surface the concern here... I'm still kind of worried about the clobbering of old blocks when they are saved with the same height. Maybe we should not make the block number the primary key, and instead just add a new column with the sole purpose of primary key. Then we can have a uniqueness constraint on block height and make sure that we handle any reorgs around it.

Reply:
Ok, noted. On the plus side, the work put into `HeadTracker` will give us the groundwork needed to handle conflicting block numbers (aka height) at a later date.